### PR TITLE
Fixes and refactoring regarding LibBytes

### PR DESCRIPTION
**Pull request by Recmo**
## Description

Fixes:
* Fix trailing garbage in `LibBytes.equals` (formerly `areBytesEqual`)
* Fix masking of unused bits in `writeAddress` and `readBytes4`.

Refactors:
* Merge LibBytes and LibMem.
* Refactor into `using LibBytes for bytes`.
* Rename a bunch of functions for clarity and consistency.
* Factor isFunctionRemoveAuthorizedAddress to test

I also added `slice` and `sliceDestructive`. The later is a generalization of the `pop*` functions. They can be used to pop from the front as easily as from the back. If we decide to use them we should add tests, currently they have none. If we don't use them they should be removed.

## Types of changes

*   [x] Bug fix (non-breaking change which fixes an issue)
*   [ ] New feature (non-breaking change which adds functionality)
*   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [ ] Change requires a change to the documentation.
*   [x] Added tests to cover my changes.
*   [ ] Added new entries to the relevant CHANGELOG.jsons.
*   [ ] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [ ] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/contracts/src/contracts/current/test/TestAssetProxyOwner/TestAssetProxyOwner.sol [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    )
+        public
+        AssetProxyOwner(_owners, _assetProxyContracts, _required, _secondsTimeLocked)
+    {
+    }
+    
+    /// @dev Compares first 4 bytes of byte array to removeAuthorizedAddress function selector.
+    /// @param data Transaction data.
+    /// @return Successful if data is a call to removeAuthorizedAddress.
+    function isFunctionRemoveAuthorizedAddress(bytes memory data)
```
[the remark this answers is not in the document]
Comment (Recmo) [reply packages/contracts/src/contracts/current/test/TestAssetProxyOwner/TestAssetProxyOwner.sol]: Hmm, I think I prefer not having testing code in the production contracts. It's also a bit of a quirky function, the name and return value suggests it returns a bool, but it actually reverts. And its a public function, so it will show up in the ABI.

We could make it internal and expose it with a public wrapper in the testing contract.

But IMHO, in this case the extra layer of indirection for what is essentially a call to readBytes4 and a comparison hurts the code quality more than what we gain from more consistent testing.

Perhaps there is another way we can test this comparison? This would test the whole modifier:

```
function testStoreTransaction(Transaction tx) returns (uint256 id) {...}

function testValidRemoveAuthorizedAddressTx(uint256 id)
   validRemoveAuthorizedAddressTx(id)
   { /* do nothing, revert through modifier */ }
```

**Comment by Recmo** [comment]
> I like the new syntax, but do we want to change this for other libraries as well?

We could do it with SafeMath `using SafeMath for uint256` would give us `a.safeMul(b)`, etc.

Others, like `LibOrder`, `LibFillResult`. I'm not sure if this syntax would help much. It would be nice to do `order.hash()`. And `totalFill.add(singleFill)`. But both are just one function. And it kind of adds an indirection. Here it seems more clear if we explicitly wrote `LibOrder.orderHash(order)` and `LibFillResult.addFillResults(total, single)`.

I do think it makes sense to turn all `Lib*` contracts into libraries instead of using inheritance. Right now finding where something is defined is a bit of a wild goose-chase. And the way inheritance overrides it is hard to tell if something is the final definition.

Just to clarify as this is not obvious: public functions in libraries will use linking and `DELEGATECALL` which is costly.
Internal functions in libraries will get included into the using contract at compile time. There's no linking or overhead involved.

File: packages/contracts/src/contracts/current/test/TestAssetProxyOwner/TestAssetProxyOwner.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (abandeali1) [reply packages/contracts/src/contracts/current/test/TestAssetProxyOwner/TestAssetProxyOwner.sol]: I like testing modifier this way.

File: packages/contracts/src/contracts/current/utils/LibBytes/LibBytes.sol [hunk 2]
```diff
@@ -31,6 +28,187 @@ contract LibBytes is
...
+        // Create a new bytes structure and copy contents
+        result = new bytes(to - from);
+        memCopy(
+            result.contentAddress(),
+            b.contentAddress() + from,
+            result.length);
+        return result;
+    }
+    
+    /// @dev Returns a slices from a byte array without preserving the input.
```
Comment (dekz) [inline packages/contracts/src/contracts/current/utils/LibBytes/LibBytes.sol]: `slice`

File: packages/contracts/src/contracts/current/utils/LibBytes/LibBytes.sol [hunk 3]
```diff
@@ -31,6 +28,187 @@ contract LibBytes is
...
+    /// @param b The byte array to take a slice from. Will be destroyed in the process.
+    /// @param from The starting index for the slice (inclusive).
+    /// @param to The final index for the slice (exclusive).
+    /// @return result The slice containing bytes at indices [from, to)
+    /// @dev When `from == 0`, the original array will match the slice. In other cases it's state will be corrupted.
```
Comment (dekz) [inline packages/contracts/src/contracts/current/utils/LibBytes/LibBytes.sol]: `its`

**Comment by Recmo** [comment]
Rebased on `v2-prototype` in preparation for merger. No conflicts.
